windows/CodePush.Net46/UpdateManager.cs

```diff
@@ -0,0 +1,326 @@
+                // metadata so that we can find and run it easily the next time.
+                string relativeBundlePath = await UpdateUtils.FindJSBundleInUpdateContentsAsync(newUpdateFolder, expectedBundleFileName).ConfigureAwait(false);
+                if (relativeBundlePath == null)
+                {
+                    throw new InvalidDataException("Update is invalid - A JS bundle file named \"" + expectedBundleFileName + "\" could not be found within the downloaded contents. Please check that you are releasing your CodePush updates using the exact same JS bundle file name that was shipped with your app's binary.");
+                }
+                else
+                {
+                    if (diffManifestFile != null)
+                    {
+                        // TODO verify hash for diff update
+                        // CodePushUpdateUtils.verifyHashForDiffUpdate(newUpdateFolderPath, newUpdateHash);
+                    }
+
+                    updatePackage[CodePushConstants.RelativeBundlePathKey] = relativeBundlePath;
+                }
+            }
+            catch (InvalidDataException)
+            {
+                // Downloaded file is not a zip, assume it is a jsbundle
+                await downloadFile.RenameAsync(expectedBundleFileName).ConfigureAwait(false);
+                await downloadFile.MoveAsync(newUpdateFolder.Path, NameCollisionOption.ReplaceExisting).ConfigureAwait(false);
+            }
+            catch (Exception e)
+            {
+                throw e;    //left here for debug purposes to see the reason of ZipFile failure.
+            }
+
+            // Save metadata to the folder
+            await newUpdateMetadataFile.WriteAllTextAsync(JsonConvert.SerializeObject(updatePackage)).ConfigureAwait(false);
+        }
+
+        internal async Task<JObject> GetCurrentPackageAsync()
+        {
+            string packageHash = await GetCurrentPackageHashAsync().ConfigureAwait(false);
+            return packageHash == null ? null : await GetPackageAsync(packageHash).ConfigureAwait(false);
+        }
+
+        internal async Task<IFile> GetCurrentPackageBundleAsync(string bundleFileName)
+        {
+            var packageFolder = await GetCurrentPackageFolderAsync().ConfigureAwait(false);
+            if (packageFolder == null)
+            {
+                return null;
+            }
+
+            JObject currentPackage = await GetCurrentPackageAsync().ConfigureAwait(false);
+            var relativeBundlePath = (string)currentPackage[CodePushConstants.RelativeBundlePathKey];
+
+            return relativeBundlePath == null
+                ? await packageFolder.GetFileAsync(bundleFileName).ConfigureAwait(false)
+                : await packageFolder.GetFileAsync(relativeBundlePath).ConfigureAwait(false);
+        }
+
+        internal async Task<string> GetCurrentPackageHashAsync()
+        {
+            JObject info = await GetCurrentPackageInfoAsync().ConfigureAwait(false);
```

Comment:
>JObject [](start = 12, length = 7)

Kind of a nitpick, but there seems to be an inconsistent use of explicit and implicit variable types.  On the whole, does the code use mostly `var` or is there already a healthy mix? The C# coding convention says "Use implicit typing for local variables when the type of the variable is obvious from the right side of the assignment, or when the precise type is not important."  I rarely find a case where type information is really important.